src/include/sof/audio/module_adapter/module/generic.h

```diff
@@ -225,6 +225,35 @@ struct processing_module {
 	/* module-specific flags for comp_verify_params() */
 	uint32_t verify_params_flags;
 
+	/* indicates that this DP module did not yet reach its first deadline and
+	 * no data should be passed yet to next LL module
+	 *
+	 * why: lets assume DP with 10ms period (a.k.a a deadline). It starts and finishes
+	 * Earlier, i.e. in 2ms providing 10ms of data. LL starts consuming data in 1ms chunks and
+	 * will drain 10ms buffer in 10ms, expecting a new portion of data on 11th ms
+	 * BUT - the DP module deadline is still 10ms, regardless if it had finished earlier
+	 * and it is completely fine that processing in next cycle takes full 10ms - as long as it
+	 * fits into the deadline.
+	 * It may lead to underruns:
+	 *
+	 * LL1 (1ms) ---> DP (10ms) -->LL2 (1ms)
+	 *
+	 * ticks 0..9 -> LL1 is producing 1ms data portions, DP is waiting, LL2 is waiting
+	 * tick 10 - DP has enough data to run, it starts processing
+	 * tick 12 - DP finishes earlier, LL2 starts consuming, LL1 is producing data
+	 * ticks 13-19 LL1 is producing data, LL2 is consuming data (both in 1ms chunks)
+	 * tick 20  - DP starts processing a new portion of 10ms data, having 10ms to finish
+	 *	      !!!! but LL2 has already consumed 8ms !!!!
+	 * tick 22 - LL2 is consuming the last 1ms data chunk
```

Comment:
As you assumption, DP need 2ms to generate 10ms data, for the second time, DP start processing on tick 20, so it should finish on tick 22, at tick 22, LL2 can get 1ms data.

assumption maybe based on DP need take 3ms to finish 10ms process.

Reply:
again, why do you think that DP will finish in 2ms at second cycle? And 3rd? and later? What if a second pipeline starts with 2 more DPs with 3ms deadlines (that will shorter deadlines and therefore will be scheduled before 10ms one)?

The only guaranteed time is a deadline

Reply:
i.e. 

we have 2DPs - one 10ms, process time 5ms, second 5ms, process time 2ms

```LL --> DP1 (10ms / 5ms) ---> LL ---> DP2 (5ms / 2ms) --> LL```

1st cycle  - DP2 is not yet processing, full CPU power to DP1
```|  DP1 1,2,3,4,5|  finish in 5ms```


later - CPU time is sliced between DP1 and DP2
```
| DP1 1,2,3|   |DP2 1, 2|  | DP1 4,5| (idle for 1ms) |DP2 1, 2|  DP1 1,2,3|   |DP2 1, 2|  | DP1 4,5|
                                 ^
                           finish in 7ms
```

CPU loaded in 90%, DP1 finishes in 7ms (deadline met) regardless it processes 5ms, DP2 finishes in 2ms (deadline met)

Reply:
it is configurable by the module itself. 
It can set any period it wants during prepare method
or it will be calculated based on OBS and data rate